# Live Code 3

## Muhammad Fahrur Rozi Butar-Butar
## FTDS-016

## Description

### Objective

1. Tampilkan tabel crime sampai 5 baris pertama
2. Apa saja jenis tindak kejahatan yang terjadi di kota Austin selama tahun 2016?
3. Berapa banyak kasus tindak kejahatan berdasarkan kasus pembersihannya (clearance status) selama tahun 2016?
4. Jenis kejahatan apa saja yang paling banyak belum terselesaikan kasusnya di tahun 2016?
5. Di bulan apa di tahun 2016 kasus pencurian (theft) sangat banyak terjadi?
6. Di distrik apa yang paling banyak terjadi kasus pencurian selama 2016?
7. Berapa lama rata-rata waktu (dalam hari) kasus tindak kejahatan 'theft' terselesaikan sejak kasus tersebut terangkat selama tahun 2016? (`Hint: Hitung selisih hari dari timestamp hingga clearance date`)

### Problem Statement

Definisikan problem statement berdasarkan penjabaran masalah di atas

### Import libray yang dibutuhkan

In [48]:
import pandas as pd

### Menghubungkan Google Collab dengan Bigquerry

In [35]:
from google.colab import auth
auth.authenticate_user()
print('Authenticated')

Authenticated


In [36]:
from google.cloud import bigquery

In [37]:
project_id = "emerald-surface-363009" 

In [38]:
client = bigquery.Client(project=project_id)

#### Untuk memudahkan analisa dan querry data maka kita terlebih dahulu harus tau missing value tiap kolom agar, querry yang dibuat akan menjadi akurat

In [39]:
df = client.query('''
SELECT *
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016
''').to_dataframe()

In [40]:
df.isnull().sum()

unique_key                   1
address                      0
census_tract               167
clearance_date            1719
clearance_status          1719
council_district_code      256
description                  0
district                     0
latitude                 37461
longitude                37461
location                 37461
location_description         0
primary_type                 0
timestamp                    0
x_coordinate               867
y_coordinate               867
year                         0
zipcode                    167
dtype: int64

Ternyata banyak missing value yang tedapat pada dataset maka dari itu kita harus berhati-hati memilih kolom yang ingin kita querry agar hasil menjadi akurat

## Penjabaran Masalah

 #### 1. Tampilkan tabel crime sampai 5 baris pertama

```
SELECT *
FROM `bigquery-public-data.austin_crime.crime` 
LIMIT 5
```

In [53]:
soal_1 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_1.csv')
soal_1

unique_key     address  census_tract                  clearance_date  \
0  20165002898  Austin, TX           NaN  2016-03-14 12:00:00.000000 UTC   
1  20165004843  Austin, TX           NaN  2016-03-31 12:00:00.000000 UTC   
2   2016250988  Austin, TX           NaN  2016-09-16 12:00:00.000000 UTC   
3  20165003839  Austin, TX           NaN  2016-02-09 12:00:00.000000 UTC   
4   2016321746  Austin, TX           NaN  2016-03-22 12:00:00.000000 UTC   

  clearance_status  council_district_code                     description  \
0      Not cleared                    NaN  RAPE                             
1      Not cleared                    NaN  RAPE                             
2      Not cleared                    NaN  RAPE                             
3      Not cleared                    NaN  AGG RAPE OF A CHILD              
4      Not cleared                    NaN  RAPE OF A CHILD                  

  district  latitude  longitude  location  location_description primary_type  \
0       UK       NaN        NaN       NaN                   NaN         Rape   
1       UK       NaN        NaN       NaN                   NaN         Rape   
2       UK       NaN        NaN       NaN                   NaN         Rape   
3       UK       NaN        NaN       NaN                   NaN         Rape   
4       UK       NaN        NaN       NaN                   NaN         Rape   

                        timestamp  x_coordinate  y_coordinate  year  zipcode  
0  2016-01-19 12:00:00.000000 UTC           NaN           NaN  2016      NaN  
1  2016-01-25 12:00:00.000000 UTC           NaN           NaN  2016      NaN  
2  2016-01-25 12:00:00.000000 UTC           NaN           NaN  2016      NaN  
3  2016-01-27 12:00:00.000000 UTC           NaN           NaN  2016      NaN  
4  2016-02-01 12:00:00.000000 UTC           NaN           NaN  2016      NaN

#### 2. Apa saja jenis tindak kejahatan yang terjadi di kota Austin selama tahun 2016?

```
SELECT DISTINCT(primary_type)
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016
```

In [54]:
soal_2 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_2.csv')
soal_2

primary_type
0         Rape
1        Theft
2      Robbery
3  Agg Assault
4     Burglary
5   Auto Theft
6       Murder

#### 3. Berapa banyak kasus tindak kejahatan berdasarkan kasus pembersihannya (clearance status) selama tahun 2016?

```
SELECT 
  clearance_status
  , COUNT(primary_type) as total_case
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016
GROUP BY clearance_status
```

In [55]:
soal_3 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_3.csv')
soal_3

clearance_status  total_case
0           Not cleared       29715
1  Cleared by Exception        1014
2                   NaN        1719
3     Cleared by Arrest        5013

Saya memilih kolom primary_type sebagai total_case karena pada kolom tersebut terdapat 0 missing value sehingga, total kasus yang didapat tidak berkurang karena adanya nilai yang kosong

#### 4. Jenis kejahatan apa saja yang paling banyak belum terselesaikan kasusnya di tahun 2016?

```
SELECT 
  primary_type
  , COUNT(primary_type) as total_case
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND clearance_status = 'Not cleared'
GROUP BY primary_type
ORDER BY total_case DESC
LIMIT 3
```

In [56]:
soal_4 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_4.csv')
soal_4

primary_type  total_case
0        Theft       21894
1     Burglary        4440
2   Auto Theft        1540

Karena di pertanyaan terdapat 'Jenis kejahatan apa saja ' maka saya menggunakan LIMIT 3, karena di soal menanyakan 'apa saja' yang berarti lebih dari 1

Saya memilih kolom primary_type sebagai total_case karena pada kolom tersebut terdapat 0 missing value sehingga, total kasus yang didapat tidak berkurang karena adanya nilai yang kosong

#### 5. Di bulan apa di tahun 2016 kasus pencurian (theft) sangat banyak terjadi?

```
SELECT 
  primary_type
  , EXTRACT(month FROM timestamp) AS month
  , COUNT(primary_type) as total_case
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND primary_type = 'Theft'
GROUP BY primary_type, month
ORDER BY total_case DESC
LIMIT 1
```

In [57]:
soal_5 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_5.csv')
soal_5

primary_type  month  total_case
0        Theft      4        2282

Karena di pertanyaan terdapat 'sangat banyak' yang berarti kasus tertinggi, sehingga saya menggunkan LIMIT 1 untuk menampilkan 1 kasus yang paling tinggi

Saya memilih kolom primary_type sebagai total_case karena pada kolom tersebut terdapat 0 missing value sehingga, total kasus yang didapat tidak berkurang karena adanya nilai yang kosong

#### 6. Di distrik apa yang paling banyak terjadi kasus pencurian selama 2016?

```
SELECT 
  district
  , COUNT(primary_type) as total_case
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND primary_type = 'Theft'
GROUP BY district
ORDER BY total_case DESC
LIMIT 1
```

In [58]:
soal_6 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_6.csv')
soal_6

district  total_case
0        D        4117

Karena di pertanyaan terdapat 'paling banyak' makan saya menampilkan 1 row yang paling tertinggi kasusnya

Saya memilih kolom primary_type sebagai total_case karena pada kolom tersebut terdapat 0 missing value sehingga, total kasus yang didapat tidak berkurang karena adanya nilai yang kosong

#### 7. Berapa lama rata-rata waktu (dalam hari) kasus tindak kejahatan 'theft' terselesaikan sejak kasus tersebut terangkat selama tahun 2016? (Hint: Hitung selisih hari dari timestamp hingga clearance date)

```
SELECT 
  ROUND(AVG(TIMESTAMP_DIFF(clearance_date, timestamp, DAY))) AS rata_rata_hari 
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND primary_type = 'Theft' AND (clearance_status = 'Cleared by Exception' OR clearance_status = 'Cleared by Arrest')
```

In [60]:
soal_7 = pd.read_csv('P0LC3_M_Fahrur_Rozi_BB_7.csv')
soal_7

rata_rata_hari
0            14.0

Pada kolom clearance_status terdapat 4 unique value yaitu Not cleared, Cleared by Exception, None, Cleared by Arrest.
Karena ditanya kasus yang 'terselesaikan' maka saya memilih 'Cleared by Exception' dan 'Cleared by Arrest'

Saya melakukan pembulatan dengan desimal 0 karena rata-rata hari tidak mungkin berkoma

## Conclusions, Assumptions, Overall Analysis

Dari penjabaran yang sudah kamu bahas (pertanyaan 2-7), apa kesimpulan yang saya dapatkan:

PADA TAHUN 2016
1. Ternyata jenis kasus kejahatan pada Austin Smart City adalah Rape, Theft, Robbery, Agg Assault, Burglary, Auto Theft, Murder.
2. Ternyata kasus tindak kejahatan berdasarkan status yang paling banyak adalah kasus yang tidak terselesaikan dengan 29715 total kasus. 
3. Jenis kasus kejahatan yang belum terselesaikan paling banyak terdapat pada kasus Theft dengan 21894 total kasus, di susul dengan Burglary dengan 4440 total kasus dan Auto Theft	sebanyak 1540 total kasus.
4. Kasus pencurian / Theft terbanyak terjadi pada bulan	4	dengan total kasus 2282 
5. kasus pencurian / theft paling banyak terjadi di district D
6. Rata-rata waktu kasus tindak kejahatan 'theft' terselesaikan sejak kasus tersebut terangkat adalah 14 hari.